lua/rustaceanvim/test.lua

```diff
@@ -1,65 +1,56 @@
 ---@return rustaceanvim.Diagnostic[]
 ---@diagnostic disable-next-line: inject-field
-M.parse_diagnostics = function(file_name, output, bufnr)
+function M.parse_cargo_test_diagnostics(output, bufnr)
   output = output:gsub('\r\n', '\n')
-  local lines = vim.split(output, '\n')
-  ---@type rustaceanvim.Diagnostic[]
   local diagnostics = {}
-  for i, line in ipairs(lines) do
-    local message = ''
-    local test_id, file, lnum, col = line:match("thread '([^']+)' panicked at ([^:]+):(%d+):(%d+):")
-    if lnum and col and message and vim.endswith(file_name, file) then
-      local next_i = i + 1
-      while #lines >= next_i and lines[next_i] ~= '' do
-        message = message .. lines[next_i] .. '\n'
-        next_i = next_i + 1
-      end
-      local diagnostic_lnum = tonumber(lnum) - 1
-      local diagnostic_col = tonumber(col) or 0
-      ---@type rustaceanvim.Diagnostic
-      local diagnostic = {
-        bufnr = bufnr,
-        test_id = test_id,
-        lnum = diagnostic_lnum,
-        end_lnum = diagnostic_lnum,
-        col = diagnostic_col,
-        end_col = diagnostic_col,
-        message = message,
-        source = 'rustaceanvim',
-        severity = vim.diagnostic.severity.ERROR,
-      }
-      table.insert(diagnostics, diagnostic)
-    end
+
+  for failure_content, test_id, lnum, col in
+    output:gmatch("(thread '([^']+)' panicked at [^:]+:(%d+):(%d+):%s-\n.-\n\n)")
+  do
+    table.insert(diagnostics, {
+      bufnr = bufnr,
+      test_id = test_id,
+      lnum = lnum,
+      end_lnum = lnum,
+      col = col,
+      end_col = col,
+      message = failure_content,
+      source = 'rustaceanvim',
+      severity = vim.diagnostic.severity.ERROR,
+    })
   end
-  if #diagnostics == 0 then
-    --- Fall back to old format
-    for test_id, message, file, lnum, col in
-      output:gmatch("thread '([^']+)' panicked at '([^']+)', ([^:]+):(%d+):(%d+)")
-    do
-      if vim.endswith(file_name, file) then
-        local diagnostic_lnum = tonumber(lnum) - 1
-        local diagnostic_col = tonumber(col) or 0
-        ---@type rustaceanvim.Diagnostic
-        local diagnostic = {
-          bufnr = bufnr,
-          test_id = test_id,
-          lnum = diagnostic_lnum,
-          end_lnum = diagnostic_lnum,
-          col = diagnostic_col,
-          end_col = diagnostic_col,
-          message = message,
-          source = 'rustaceanvim',
-          severity = vim.diagnostic.severity.ERROR,
-        }
-        table.insert(diagnostics, diagnostic)
-      end
-    end
+
+  return diagnostics
+end
+
+---@param junit_xml string
+---@param bufnr integer
+---@return rustaceanvim.Diagnostic[]
+---@diagnostic disable-next-line: inject-field
+function M.parse_nextest_diagnostics(junit_xml, bufnr)
+  local diagnostics = {}
+  for test_id, _, lnum, col, failure_content in
+    junit_xml:gmatch(
+      '<failure.-message="thread &apos;([^;]+)&apos; panicked at ([^:]+):(%d+):(%d+)".-<system%-err>(.-)</system%-err>'
+    )
```

Comment:
note: I suppose we have to parse the xml manually here, as this function may be used outside of the neotest adapter (and neotest may not be installed).

It may be beneficial to have a separate version of this function in the `rustaceanvim.neotest` module that is used by the neotest adapter only, and uses the neotest xml library.

Reply:
kinda related to [this comment](https://github.com/mrcjkb/rustaceanvim/pull/838#discussion_r2280964295), but what's the point of doing that? 😅